### ISSUE TITLE
Jackson2HalModule stack trace de-serializing w/ Resource class

### ISSUE THREAD

**Issue by keaplogik** [spring-projects/spring-hateoas#325]
When attempting to unwrap the content of a resource using the hal module, we run into a stack trace de-serializing the model. It seems to have issues with the Link classes templated field. 

Example:

``` java
class User{
    private String name;
    public String getName(){ return this.name; }
    public String setName(String name){ this.name = name; }
}

//Register HAL Deserializer.
objectMapper.registerModule(new Jackson2HalModule());
//objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
objectMapper.setHandlerInstantiator(
        new Jackson2HalModule.HalHandlerInstantiator(new AnnotationRelProvider(), null)
);

Resources<Resource<User> userResources = this.objectMapper.readValue(
        "{
            '_links': { 
                'self': { 
                    'href': '../users'
                },
               'findByName': {
                   'href': '../users/findByName{?name}',
                   'templated': true
               }
            },
            '_embedded': { 
                'users': [ 
                    {  
                        '_links': {
                            'self': { 
                                'href': '../users/1'
                            } 
                        }, 
                        'name': 'New Name'
                    }
                ]
            }
          }",
        this.objectMapper.getTypeFactory().constructParametricType(Resources.class,
                this.objectMapper.getTypeFactory().constructParametricType(
                        Resource.class, User.class)));
```

```
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "templated" (class org.springframework.hateoas.Link), not marked as ignorable (one known property: "href"])
 at [Source: java.io.StringReader@388ba540; line: 1, column: 123] (through reference chain: org.springframework.hateoas.ResourceSupport["_links"]->org.springframework.hateoas.Link["templated"])
    at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:51)
    at com.fasterxml.jackson.databind.DeserializationContext.reportUnknownProperty(DeserializationContext.java:671)
    at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:773)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1297)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1275)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:247)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
    at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:2965)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1587)
    at com.fasterxml.jackson.core.JsonParser.readValueAs(JsonParser.java:1431)
    at org.springframework.hateoas.hal.Jackson2HalModule$HalLinkListDeserializer.deserialize(Jackson2HalModule.java:500)
    at org.springframework.hateoas.hal.Jackson2HalModule$HalLinkListDeserializer.deserialize(Jackson2HalModule.java:446)
    at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:525)
    at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:106)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:242)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
    at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:2993)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2098)
```

**Comment by keaplogik** [comment]
I don't think this LinkMixin is needed since the Link will reconstruct itself correctly and isTemplated is determined by the Link's backed UriTemplate.

See line https://github.com/spring-projects/spring-hateoas/blob/master/src/main/java/org/springframework/hateoas/hal/LinkMixin.java#L44

**Comment by fcbarnes15** [comment]
To test this out further,  I modified the renderSingleLinkAsObject() and deserializeSingleLink() tests in Jackson2HalIntegrationTest.java to serialize and deserialize a Link that has a search param in it. It successfully serializes with "templated":true.  But you can't deserialize it back.  I went further and removed the "templated":true from the json string and I can deserialize correctly and the Link.isTemplated() returns true as you would expect.  So the question becomes why is templated=true being preserved in the serialized representation if it isn't needed to reconstruct the object?

**Comment by olivergierke** [comment]
What's the reason you're not setting `FAIL_ON_UNKNOWN_PROPERTIES` to false? Using Jackson's default will basically blow up inevitably as it rejects all content not mapped 1:1 on the server.

Generally speaking, sending HAL to a server is a bit weird in the first place as it's a read-only format. What you actually want to send is:

```json
{
  "name" : …
}
```

which – as you can see – is way more concise.

**Comment by keaplogik** [comment]
We are using `HalObjectMapper` on the clients to deserialize our responses, and provide client tooling for traversion in a loadbalanced MSA. We default the mapper `FAIL_ON_UNKNOWN_PROPERTIES ` to false which I think is fine. Still not sure why the templated is preserved, but we are fine with closing this story if you except this a non issue

Thanks.

**Comment by ledor473** [comment]
As stated by @fcbarnes15, if this is serialize it should be able to deserialize it.
I totally agree that it's a good practice to have `FAIL_ON_UNKNOWN` to false, but it's not a requirement.

**Comment by gregturn** [comment]
Resolved via #667.
